Sources/Controllers/MyPlaces/OAFavoriteListViewController.mm

```diff
@@ -294,114 +295,131 @@ -(CGFloat) getToolBarHeight
 
 - (void)updateDistanceAndDirection
 {
-    [self updateDistanceAndDirection:NO];
+    dispatch_async(dispatch_get_main_queue(), ^{
+        [self updateDistanceAndDirection:NO];
+    });
 }
 
 - (void)updateDistanceAndDirection:(BOOL)forceUpdate
 {
-    dispatch_async(dispatch_get_main_queue(), ^{
-        if ([self.favoriteTableView isEditing])
-            return;
-
-        if ([[NSDate date] timeIntervalSince1970] - self.lastUpdate < 0.3 && !forceUpdate)
-            return;
-        self.lastUpdate = [[NSDate date] timeIntervalSince1970];
-
-        OsmAndAppInstance app = [OsmAndApp instance];
-        // Obtain fresh location and heading
-        CLLocation* newLocation = app.locationServices.lastKnownLocation;
-        if (!newLocation)
+    if ([self.favoriteTableView isEditing])
+        return;
+    
+    CFTimeInterval currentTime = CACurrentMediaTime();
+    if (currentTime - self.lastUpdate < 0.3 && !forceUpdate)
+        return;
+    self.lastUpdate = currentTime;
+    
+    OsmAndAppInstance app = [OsmAndApp instance];
+    // Obtain fresh location and heading
+    CLLocation *newLocation = app.locationServices.lastKnownLocation;
+    if (!newLocation)
+        return;
+    
+    CLLocationDirection newHeading = app.locationServices.lastKnownHeading;
+    CLLocationDirection newDirection =
+    (newLocation.speed >= 1 /* 3.7 km/h */ && newLocation.course >= 0.0f)
+    ? newLocation.course
+    : newHeading;
+    
+    __weak __typeof(self) weakSelf = self;
+    dispatch_barrier_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
```

Comment:
it is not okay to use dispatch_barrier_async with any of the global concurrent queues obtained using dispatch_get_global_queue